### ISSUE TITLE
Rule: S2189 detect infinite loops

### ISSUE THREAD

**Issue by fmallet** [SonarSource/sonar-dotnet#199]
Should implement [RSPEC-2189](https://jira.sonarsource.com/browse/RSPEC-2189)

**Comment by Corniel** [comment]
I had a short look into this, taking [RSPEC-2189](https://jira.sonarsource.com/browse/RSPEC-2189) as a starting point.

``` C#
class S2189
{
    void For() // Noncompliant 2190
    {
        for (; ; )
        {  // Noncompliant; end condition omitted
            // ...
        }
    }

    void While() // Noncompliant 2190
    {
        int j = 0;
        while (true)
        { // Noncompliant; end condition omitted
            j++;
        }
    }

    void Other()
    { 
        int k = 0;
        bool b = true;
        while (b) // Noncompliant S2589
        { // Noncompliant; b never written to in loop
            k++;
        }
    }

    IEnumerable<string> Yield()  // Noncompliant 2190
    {
        while (true)
        {
            yield return "test";
        }
    }
}
```

So, all issues are - indirect - covered by others. Is there value for implementing S2189 too? Though could argue that 2190 should not report on these, as no recursion (calling to self) is involved.
